rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+
+However, since the size of MemTable is limited and might be flushed & rotated at any time, it cannot buffer ALL the changes since the earliest running transaction. To deal with this limitation, RocksDB's solution is simply abort the transaction as 'Expired' if the MemTable's earliest sequence number is bigger than the transaction's sequence number.
+
+The good part of this approach is that it's very efficient and simple, and it's free of floating garbage. The bad part is that it may abort the transaction when it's not necessary, and tends to encourage users to set a bigger MemTable size to reduce the probability of the expired transaction.
+
+### Badger
+
+Badger takes a different approach from RocksDB, it provides Serializable isolation level by default. The key idea is to track both read and write operations in the transaction state, and check conflicts with the recent committed transactions tracked in a global Oracle during commit.
+
+Badger maintains a global "Oracle" which tracks the recent committed transactions. When a transaction commits, it should push itself into the Oracle's recent committed transactions list. The Oracle also maintains a sequence number generator, which is used to assign sequence numbers to transactions.
+
+The conflict checking in Badger is as follows:
+
+1. Track the sequence number when transaction started.
+2. Track both read and write keys in the transaction state.
+3. When `commit()` is called, check conflicts with the recent committed transactions in the Oracle which committed after the transaction started.
+4. The conflict check verifies that: the transaction's read and write keys do not conflict with other transactions' write keys
+
+This approach provides Serializable isolation level by detecting both Write-Write conflicts and Read-Write conflicts. It does not rely on MemTable size like RocksDB does, and provides more accurate conflict detection, without worrying about the MemTable size.
+
+The downside is that it requires more memory to track the read keys and recent committed transactions. However, Badger mitigates this by:
+
+1. Using u64 key fingerprints instead of full keys to track read/write keys.
+2. Having an GC mechanism to clean up old committed transactions from the Oracle.
+
+The GC is also straightforward - when all running transactions have sequence numbers greater than a committed transaction's sequence number, that committed transaction can be removed from the Oracle.
+
+## Proposal
+
+### API
+
+The transaction API should be resembles with the non-transactioned API. All the api like `get()`, `put()`, `range()` should also be supported in the transaction API, and these APIs should be aligned with the Snapshot API + WriteBatch API. The only difference is that the transaction API should be able to `commit()` or `rollback()` the changes.
```

Comment:
With transaction tracking, range scans could get rather expensive 🤔 as a followup, would it make sense to think about a more optimal data structure for bookkeeping range scans, than simply tracking every single read key?

Reply:
Ah, nvm ... it seems to be covered already

```
    read_keys: Vec<Range<Bytes>>,
```